API 2.0

kahaaga opened the pull request:
This is the cleaned-up version of #97 (which I am keeping temporarily, to have access to timings). When tests pass and documentation looks ok, we can merge this and delete #97.

Datseris commented:
Yeah, but now things become a bit difficult. This PR introduces 600 loc changes if not more. Not sure how possible it is to review it for me, it feels as if every line of code was changed. I wonder if such a massive change was really necessary though?

In any case, how did you ensure that the surrogate generate before and after this change remains "correct"? I now see more and more how seriously we need to take #66 coz now the test tells us pretty much nothing whether something broke in this PR or not xD

Datseris commented on src/methods/aaft.jl:
```diff
@@ -16,14 +17,76 @@ monotonic nonlinear transformation of a linear Gaussian process
+    # Updated Fourier transform, with shuffled phases
+    shuffled𝓕 .= r .* exp.(coeffs .* 1im)
+    
+    # Inverse Fourier transform
+    s .= (inverse * shuffled𝓕) .+ m
```
Personally, I'm not a fan of comments like these. And in fact, the same opinion is stated in books that teach how to write elegant code. Comments like these provides no real information. They tell you what the line of code says by itself. So the comments inflate the total amount of code and make it harder to review and navigate, but its not like they help the user understand by providing more information. 

Example:
```julia
# square y and then get logarithm for the answer
answer = log(y^2)
```

Another thing that makes the code difficult to read is that it has too many breaks in empty lines. Similar to a paper, contiguous processes form a paragraph. They don't have empty lines that interrupt them. Instead, empty lines separate different paragraphs. 

If you have some spare time, have a look in this workshop I gave about writing good code: https://youtu.be/iIIBFujl254 it uses information from various sources and own experience.

kahaaga replied:
Thanks for the link! I'll at least remove redundant code from the PR. 

> _"proper use of comments is to compensate for our failure  to express ourselves in code"_

Oh yes, I feel that. 😁

Datseris replied:
hahahahahahah yeah I know!!!!

kahaaga commented:
> In any case, how did you ensure that the surrogate generate before and after this change remains "correct"? I now see more and more how seriously we need to take #66 coz now the test tells us pretty much nothing whether something broke in this PR or not xD

I've tested individually (visually, of course, since we have no formal way of testing the correctness) for every method as I went along. For the final PR, my idea was that if the surrogates in the documentation look reasonable (i.e. as before), and the test pass (which ensures that basic properties of the surrogates at least are correct), then we can go ahead and merge.

> Yeah, but now things become a bit difficult. This PR introduces 600 loc changes if not more. Not sure how possible it is to review it for me, it feels as if every line of code was changed. I wonder if such a massive change was really necessary though?

Yes, that is *kind of* awkward. Not sure how to proceed. Maybe most of this stuff should have gone in individual PRs.
Any ideas?

Datseris commented:
> I've tested individually (visually, of course, since we have no formal way of testing the correctness) for every method as I went along. For the final PR, my idea was that if the surrogates in the documentation look reasonable (i.e. as before), and the test pass (which ensures that basic properties of the surrogates at least are correct), then we can go ahead and merge.

That is good enough for me actually :D #66 is a monster to tackle by itself, I am certainly not expecting to tackle it before we submit the paper.

Datseris commented:
On a different note, 
```
s .= (inverse * shuffled𝓕) .+ m
```
allocates one extra array `(inverse * shuffled𝓕)`. There aren't in-place fourier transform versions?

kahaaga commented:
> It isn't possible to do inverse .* shuffled ?

Unfortunately not yet. For the forward transform, we can use `LinearAlgebra.mul!(𝓕, forward, s)` instead of `𝓕 .= forward * s`, but this doesn't work for the inverse transform. Hopefully, the inverse in-place transform will be merged in FFTW.jl soon, so we can do that (I also asked about the status of the FFTW.jl PR addressing that today).

Datseris commented:
It's okay, let's not worry about the inverse now, we get really good speedups throughout anyways.

kahaaga commented:
@Datseris Are you able to open the Github merge conflict tool for the `src/methods/randomfourier.jl` file? I can't at my end. We need this conflict to be fixed in order to trigger CI.

Anyways, I'll stop commenting and changing things until you're done having a look, so we don't fix things twice.

Datseris commented:
I also couldn't, but I could do it locally and push it here! I've marked this ready for review because that's also necessary for doc build I think.

Datseris commented on src/methods/lombscargle.jl:
```diff
@@ -120,8 +120,7 @@ function (sg::SurrogateGenerator{<:IrregularLombScargle})()
 function swap_elements!(candidate, s, rng, range)
-    k = sample(rng, range)
-    l = sample(rng, range)
+    k, l = sample(rng, range, 2, replace = false)
```
It is actually better to use teh previous version for performance. `sample` here allocates a 2-element vector :)

Datseris replied:
```
julia> x = rand(100);

julia> sample1(x) = sample(x, 2; replace = false);

julia> sample2(x) = (a = sample(x); b = sample(x); return (a, b))
sample2 (generic function with 1 method)

julia> sample1(x)
2-element Vector{Float64}:
 0.12229836237241765
 0.0943741108705125

julia> sample2(x)
(0.9856096085646662, 0.4488287731030546)

julia> using BenchmarkTools

julia> @btime sample1($x);
  41.532 ns (1 allocation: 80 bytes)

julia> @btime sample2($x);
  17.836 ns (0 allocations: 0 bytes)
```

kahaaga replied:
Good catch. I wanted to sample both at the same time, so I could make sure I sampled without replacement (so that we don't swap the same elements), but that isn't necessary, because it will occur very rarely, and won't affect the result of the algorithm.

kahaaga commented:
All the examples in the documentation seem to do what I'd expect them to do, so this is good to merge from my side. Again, it would be helpful to have some objective test 😬 

Let me know when you've had a look at everything, and I will remove redundant comments and fix potential other issues you bring up as soon as possible.

Datseris commented:
Just tag me when you want a final review here!

kahaaga commented:
@Datseris I've removed redundant comments and tried to style the code in a more paragraph-like style. Not all comments were removed, because I still think they're useful, but comments are not in place during every step anymore.

I also think the `IrregularLombScargle` example works fine now. The mistake was that I adjusted the number of maximum iteration too low, so that the Lomb-Scargle periodogram of the surrogate wouldn't converge sufficiently to the periodogram of the original surrogate. This is now fixed.

Datseris commented on src/api.jl:
```diff
@@ -3,28 +3,34 @@ export surrogate, surrogenerator, Surrogate
 
-struct SurrogateGenerator{S<:Surrogate, X, A, R<:AbstractRNG}
+struct SurrogateGenerator{S<:Surrogate, Xₓ, Xₛ, A, R<:AbstractRNG}
     method::S # method with its input parameters
-    x::X      # input timeseries
+    x::Xₓ      # input timeseries
+    s::Xₛ      # surrogate (usually same type as `x`, but not always)
```
Unless there is an established symbol in the scientific literature, which isn't the case here, we should use as simple and clear to read symbols as possible.

kahaaga replied:
`S` is already used for the surrogate, that's why I introduced `Xₓ` and `Xₛ`. But the symbols don't really matter here, so I'll just give it another symbol. Maybe `X` for the original time series and `Y` for the surrogate time series.

Datseris commented on src/methods/large_shuffle.jl:
```diff
@@ -19,74 +33,99 @@ Hence, these surrogates can be used to test any null hypothesis aimed at
+    blocklengths = get_uniform_blocklengths(length(x), bs.n)
+
+    # The start index of each block.
+    startinds = startinds = [1; cumsum(blocklengths) .+ 1]
```
This is more like a remark not a comment to be addressed. I think it would be more intuitive if you wrote this as `vcat(1, cumsum...)`. The reason being that brackets and `;` are used in matrix construction `[1 2; 3 4]`. But here you are actually making a vector. I first thought "he's making an array, but why?" but took me a bit of time to realize its a vector. With `vcat` there is no ambiguity really.

(p.s.: an alternative to vcat would be `pushfirst!(cumsum(blocklengths) .+ 1, 1)`

kahaaga replied:
Sure, we can go for `vcat(1, cumsum...)`. I agree that it makes it clearer that this is a vector in this context.

Datseris replied:
to be clear: `vcat(1, cumsum(blocklengths) .+ 1)`.

Datseris commented on src/methods/pseudoperiodic.jl:
```diff
@@ -40,40 +41,39 @@ function surrogenerator(x::AbstractVector, pp::PseudoPeriodic, rng = Random.defa
         y[i+1] = z[j+1]
     end
-    return y[:, 1]
+    s .= y[:, 1]
```
```suggestion
    for (i, p) in enumerate(y); s[i] = p[1]; end
```
This eliminates creating an intermediate vector via `y[:, 1]`.

Datseris commented on src/methods/trend_based.jl:
```diff
@@ -1,6 +1,20 @@
+
+function linear_trend(x)
+    l = linreg(0.0:1.0:length(x)-1.0 |> collect, x)
+    trendᵢ(xᵢ) = l[1] + l[2] * xᵢ
+    return trendᵢ.(x)
```
```suggestion
    return [l[1] + l[2]*a for a in x]
```

Datseris commented on src/methods/trend_based.jl:
```diff
@@ -34,78 +48,73 @@ struct TFTDRandomFourier <: Surrogate
-    𝓕new .= rx .* exp.(ϕs .* 1im)
+    𝓕 .= rx .* exp.(ϕs .* 1im)
+
+    # Unfortunately, we can't do inverse transform in-place yet, but 
```
```suggestion
    # TODO: Unfortunately, we can't do inverse transform in-place yet, but 
```

(just so we can find this again in the future)

Datseris commented on src/methods/truncated_fourier.jl:
```diff
@@ -50,68 +50,61 @@ function surrogenerator(x, method::TFTS, rng = Random.default_rng())
     # Surrogate starts out as a random permutation of x
-    s = x[StatsBase.sample(sg.rng, 1:L, L; replace = false)]
-    𝓕s .= forward*s
+    s .= x[StatsBase.sample(sg.rng, 1:L, L; replace = false)]
```
Since `sample` has an inplace version `sample!`  we could pre-allocate a container for the sampled indices.

kahaaga replied:
Good catch. This reduced allocations quite a bit, and also made me realize there was another low-hanging allocation optimization right next to it. Thanks!

Datseris commented on src/methods/truncated_fourier.jl:
```diff
@@ -136,13 +129,56 @@ struct TAAFT <: Surrogate
+        # we are only randomizing the high-frequency components of the
+        # signal.
+        n_preserve = ceil(Int, abs(fϵ * n))
+        ϕs[1:n_preserve] .= ϕx[1:n_preserve]
```
```suggestion
        ϕs[1:n_preserve] .= @view ϕx[1:n_preserve]
```

Datseris commented on src/methods/truncated_fourier.jl:
```diff
@@ -136,13 +129,56 @@ struct TAAFT <: Surrogate
+    elseif fϵ < 0
+        # Do the exact opposite to preserve high-frequencies
+        n_preserve = ceil(Int, abs(fϵ * n))
+        ϕs[end-n_preserve+1:end] .= ϕx[end-n_preserve+1:end]
```
```suggestion
        ϕs[end-n_preserve+1:end] .= @view ϕx[end-n_preserve+1:end]
```

Datseris commented on src/utils/interpolation.jl:
```diff
@@ -1,3 +1,9 @@
+
+
+getrange(t, n) = LinRange(minimum(t), maximum(t), n)
```
```suggestion
getrange(t, n) = range(minimum(t); stop = maximum(t), length = n)
```
It is recommended to use the `range` function instead of direct constructors of ranges.

Datseris commented on src/utils/powerspectrum.jl:
```diff
@@ -0,0 +1,83 @@
+using LinearAlgebra: mul!
+
+"""
+    prepare_spectrum(s) → Vector{Float64}
```
```suggestion
    prepare_spectrum(s, plan) → y::Vector
```

Datseris reviewed (approved):
@kahaaga wow, now that's one huge amount of work, congrats! I think I went through all of the PR. I've left only really small comments that you can easily address. Feel free to merge after you have addressed the comments!

kahaaga commented on src/api.jl:
```diff
@@ -3,28 +3,34 @@ export surrogate, surrogenerator, Surrogate
 for i in 1:1000
-    s = sg()
+    sg()
     # do stuff with s and or x
-    result[i] = stuff
+    result[i] = stuff(sg.s)
```
The api didn't change, just the example. But good point. We should keep the example as before.
